pandas/core/indexing.py

```diff
@@ -1994,21 +1994,19 @@ def _setitem_single_column(self, loc: int, value, plane_indexer) -> None:
             self.obj._clear_item_cache()
             return
 
+        self.obj._iset_item(loc, value)
+
         # We will not operate in-place, but will attempt to in the future.
         #  To determine whether we need to issue a FutureWarning, see if the
         #  setting in-place would work, i.e. behavior will change.
-        if isinstance(value, ABCSeries):
-            warn = can_hold_element(orig_values, value._values)
-        else:
-            warn = can_hold_element(orig_values, value)
+
+        new_values = self.obj._get_column_array(loc)
+        warn = can_hold_element(orig_values, new_values)
```

Comment:
with the check moved here, we can just roll this into the condition on L2011-L2018 (OK for separate PR)

Reply:
I shortened it a bit. Combining both conditions makes them significantly harder to read imo since we need an or in there which would require new brackets. The 2-step condition is a bit simpler imo